### ISSUE TITLE
Preparation for immutable objects

### ISSUE THREAD

**Issue by shadowhand** [rdlowrey/auryn#116]
Right now it is not possible to modify an immutable object using `prepare`. For instance, I might want to do this (as a contrived example):

```php
$injector->prepare('Zend\Diactoros\Response', function (&$response) {
    $response = $response->withHeader('X-API-Version', 'v2.1');
});
```

Attempting this will currently produce the following error:

`Warning: Parameter 1 to {closure}() expected to be a reference, value given`

Because reflection is used to execute the closure, it is not possible to use references.

What is the best way to handle immutable preparation?

**Comment by J7mbo** [comment]
Hmm, looks like you're trying to tell the injector to, when it encounters an instance of `Zend\Diactoros\Response`, inject an altered version of it (with the header). Have you considered [instantation delegation](https://github.com/rdlowrey/auryn#instantiation-delegates) for this (basically, a factory).

This way your implementation doesn't know about the injector.

    $apiResponseFactory = function() use ($response) {
        return $response->withHeader('X-API-Version', 'v2.1');
    };

    $injector->delegate('Zend\Diactoros\Response', $apiResponseFactory);

Just a suggestion. This might not work for you as it'll need to be in the same context as where you get the `$response` object.

An alternative would be to grab the injector in the `$apiResponseFactory` and use that to change the already shared `$response`. Or maybe that's complete rubbish! Perhaps @rdlowrey can chime in here :-)

**Comment by rdlowrey** [comment]
This is a sane use-case. There are a couple of options to "fix" `prepare()` so that this works ... I think the easiest is just to have it check the return value of the prepare callable with an `instanceof` and if it matches the expected class type internally replace the old instance with the newly returned instance.

This would be safe for BC as the return value from prepares wasn't previously meaningful. Can anyone think of any drawbacks to that approach that I may be missing?

**Comment by shadowhand** [comment]
@rdlowrey I like that solution very much.

**Comment by rdlowrey** [comment]
Okay, great. I'll try to implement/test/tag that change today. Will post updates here as they happen.

**Comment by J7mbo** [comment]
The `prepare()` method change looks much better, I really like that :+1:

**Comment by rdlowrey** [comment]
I've pushed the mentioned change up to master. @shadowhand if you can confirm that works as you expect (you can probably just verify by looking at the new test cases) I will go ahead and tag 1.1.0

**Comment by rdlowrey** [comment]
This capability is available as of [1.1.0](https://github.com/rdlowrey/auryn/releases/tag/1.1.0)

**Comment by kelunik** [comment]
:+1:
